plugins/wacom-usb/fu-wac-module-bluetooth-id9.c

```diff
@@ -0,0 +1,264 @@
+	fu_struct_loader_cmd_set_size(loader_cmd, bufsz + FU_STRUCT_SPI_CMD_SIZE);
+	/* CRC(concat(spi_cmd, *data)) */
+	crc = fu_crc32_full(spi_cmd->data,
+			    FU_STRUCT_SPI_CMD_SIZE,
+			    ~0,
+			    FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	crc = fu_crc32_full(buf, bufsz, ~crc, FU_WAC_MODULE_BLUETOOTH_ID9_CRC_POLYNOMIAL);
+	fu_struct_loader_cmd_set_crc(loader_cmd, crc);
+	if (!fu_struct_loader_cmd_set_data(loader_cmd, spi_cmd, error))
+		return NULL;
+
+	if (!fu_struct_loader_cmd_validate(loader_cmd->data, loader_cmd->len, 0, error))
+		return NULL;
+
+	return fu_chunk_bytes_new(g_bytes_new(loader_cmd->data, loader_cmd->len));
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_block(FuWacModule *self,
+					guint8 phase,
+					FuChunk *chunk,
+					FuProgress *progress,
+					GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GBytes) blob_chunk = NULL;
+
+	fu_byte_array_append_uint8(buf, phase);
+	g_byte_array_append(buf, fu_chunk_get_data(chunk), fu_chunk_get_data_sz(chunk));
+
+	blob_chunk = g_bytes_new(buf->data, buf->len);
+	return fu_wac_module_set_feature(self,
+					 FU_WAC_MODULE_COMMAND_DATA,
+					 blob_chunk,
+					 fu_progress_get_child(progress),
+					 FU_WAC_MODULE_WRITE_TIMEOUT,
+					 error);
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_blocks(FuWacModule *self,
+					 guint8 phase,
+					 GBytes *data,
+					 gsize block_len,
+					 FuProgress *progress,
+					 GError **error)
+{
+	g_autoptr(GPtrArray) chunks = fu_chunk_array_new_from_bytes(data, 0, 0, block_len);
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (gsize i = 0; i < chunks->len; i++) {
+		if (!fu_wac_module_bluetooth_id9_write_block(self,
+							     phase,
+							     g_ptr_array_index(chunks, i),
+							     progress,
+							     error))
+			return FALSE;
+
+		fu_progress_step_done(progress);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_extract_firmware(FuFirmware *firmware,
+					     GBytes **loader,
+					     GBytes **payload,
+					     GError **error)
+{
+	g_autoptr(GBytes) fw = fu_firmware_get_bytes(firmware, error);
+	const guint8 *blob;
+	gsize blob_len = 0;
+	gsize loader_len = 0;
+	gsize payload_len = 0;
+
+	if (fw == NULL)
+		return FALSE;
+
+	/* The firmware file is formatted as a 2 byte "length" field
+	 * followed by <length> bytes of loader code. The remainder
+	 * payload is firmware data.
+	 */
+	blob = g_bytes_get_data(fw, &blob_len);
+	if (blob_len < 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid firmware size");
+		return FALSE;
+	}
+
+	loader_len = fu_memread_uint16(blob, G_BIG_ENDIAN);
+	if (loader_len > blob_len - 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid firmware loader size");
+		return FALSE;
+	}
+	*loader = fu_bytes_new_offset(fw, 2, loader_len, NULL);
```

Comment:
you need to use error and check the return value here.

Reply:
Ack